browser/ui/webui/brave_new_tab/update_observer.cc

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/ui/webui/brave_new_tab/update_observer.h"
+
+#include <utility>
+
+#include "brave/browser/ntp_background/ntp_background_prefs.h"
+#include "brave/components/ntp_background_images/common/pref_names.h"
+
+namespace brave_new_tab {
+
+UpdateObserver::UpdateObserver(PrefService* pref_service) {
+  CHECK(pref_service);
+
+  pref_change_registrar_.Init(pref_service);
+  AddPrefListener(ntp_background_images::prefs::kNewTabPageShowBackgroundImage,
+                  Source::kBackgroundPrefs);
+  AddPrefListener(ntp_background_images::prefs::
+                      kNewTabPageShowSponsoredImagesBackgroundImage,
+                  Source::kBackgroundPrefs);
+  AddPrefListener(NTPBackgroundPrefs::kPrefName, Source::kBackgroundPrefs);
+  AddPrefListener(NTPBackgroundPrefs::kCustomImageListPrefName,
+                  Source::kBackgroundPrefs);
+}
+
+UpdateObserver::~UpdateObserver() = default;
+
+void UpdateObserver::SetCallback(
+    base::RepeatingCallback<void(Source)> callback) {
+  callback_ = std::move(callback);
+}
+
+void UpdateObserver::OnUpdate(Source update_source) {
+  if (callback_) {
+    callback_.Run(update_source);
+  }
+}
+
+void UpdateObserver::OnPrefChanged(Source update_kind,
+                                   const std::string& path) {
+  OnUpdate(update_kind);
+}
+
+void UpdateObserver::AddPrefListener(const std::string& path,
+                                     Source update_source) {
+  pref_change_registrar_.Add(
```

Comment:
This doesn't really give us a way to add more granular updates (i.e. when we add another pref source)

Reply:
See [here](https://github.com/brave/brave-core/pull/26685/files#diff-57ed566e1e4654b26baeaeb6aee43d93f7dcd10debb48867b01f72f3a12f462bR30) for an example of listening to search prefs. Does that work?

Reply:
It does indeed! I like it, its kind of like having a channel that all related notifications fall into.